04-onion-routing.md

```diff
@@ -376,6 +416,188 @@ otherwise meets the amount criterion (eg. some other failure, or
 invoice timeout), however if it were to fulfill only some of them,
 intermediary nodes could simply claim the remaining ones.
 
+### Route Blinding
+
+Nodes receiving onion packets may hide their identity from senders by
+"blinding" an arbitrary amount of hops at the end of an onion path.
+
+When using route blinding, nodes find a route to themselves from a given
+"introduction node". They then use ECDH with each node in that route to create
+a "blinded" node ID and an encrypted blob (`encrypted_data`) for each one of
+the blinded nodes.
+
+They communicate this blinded route and the encrypted blobs to the sender.
+The sender finds a route to the introduction node and extends it with the
+blinded route provided by the recipient. The sender includes the encrypted
+blobs in the corresponding onion payloads: they allow nodes in the blinded
+part of the route to "unblind" the next node and correctly forward the packet.
+
+The `encrypted_data` is a TLV stream, encrypted for a given blinded node, that
+may contain the following TLV fields:
+
+1. `tlv_stream`: `encrypted_data_tlv`
+2. types:
+    1. type: 1 (`padding`)
+    2. data:
+        * [`...*byte`:`padding`]
+    1. type: 2 (`short_channel_id`)
+    2. data:
+        * [`short_channel_id`:`short_channel_id`]
+    1. type: 4 (`next_node_id`)
+    2. data:
+        * [`point`:`node_id`]
+    1. type: 6 (`path_id`)
+    2. data:
+        * [`...*byte`:`data`]
+    1. type: 8 (`next_blinding_override`)
+    2. data:
+        * [`point`:`blinding`]
+    1. type: 10 (`payment_relay`)
+    2. data:
+        * [`u32`:`fee_base_msat`]
```

Comment:
Question to reviewers: do we want to get rid of that field? It looks like we're ready to drop the base fee from fee score cards, so if our general direction is to get rid of base fees we might as well get rid of it here?

Reply:
We should keep it, the network currently uses base fees and we want blinded routes to be compatible with that. If we deprecate base fees in the future, it will be very easy to deprecate this field too.

Reply:
You're probably right, we don't want to delay blinded paths because of this and we can easily switch to a `payment_relay_v2` without base fees whenever the network is ready.

Reply:
Good idea, let's do that! Updated in https://github.com/lightning/bolts/pull/765/commits/03f0c7d6a1834a4392c3027c796df2c55ccef72b

Reply:
The problem is that we are now leaking `log(fee_base_msat)` unless we add padding, which defeats the purpose of this optimization.